SCAutolib/controller.py

```diff
@@ -0,0 +1,419 @@
+                msg = "Can't proceed in configuration of IPA user because no " \
+                      "IPA Client is configured"
+                raise SCAutolibException(msg)
+            new_user = user.IPAUser(ipa_server=self.ipa_ca,
+                                    username=user_dict["name"],
+                                    pin=user_dict["pin"],
+                                    password=user_dict["passwd"],
+                                    card_dir=card_dir,
+                                    cert=user_dict["cert"],
+                                    key=user_dict["key"],
+                                    local=False)
+            if force:
+                new_user.delete_user()
+                user_dict["card_dir"].mkdir(exist_ok=True)
+
+            new_user.add_user()
+
+        new_card = None
+
+        if user_dict["card_type"] == "virtual":
+            hsm_conf = file.SoftHSM2Conf(
+                new_user.card_dir.joinpath("sofhtsm2.conf"),
+                new_user.card_dir)
+            hsm_conf.create()
+            hsm_conf.save()
+
+            new_card = card.VirtualCard(new_user)
+            new_card.softhsm2_conf = hsm_conf
+        else:
+            raise NotImplementedError("Other card type than 'virtual' does not "
+                                      "supported yet")
+        new_card.create()
+        new_user.card = new_card
+        self.users.append(new_user)
+
+        dump_to_json(new_user)
+
+        return new_user
+
+    def enroll_card(self, user_: user.User):
+        """
+        Enroll the card of a given user with configured CA. If private key
+        and/or the certificate are not exists, new one's would be requested
+        from corresponding CA.
+
+        :param user_: User with a card to be enrolled.
+        """
+        logger.debug(f"Starting enrollment of the card for user "
+                     f"{user_.username}")
+        if not user_.card:
+            raise SCAutolibException(f"Card for the user {user_.username} does "
+                                     f"not initialized")
+        if user_.cert is None:
+            # Creating a new private key makes sense only if the certificate
+            # doesn't exist yet
+            if user_.key is None:
+                user_.key = user_.card_dir.joinpath("private.key")
+                _gen_private_key(user_.key)
+
+            csr = user_.gen_csr()
+            if user_.cert is None:
+                user_.cert = user_.card_dir.joinpath(
+                    f"cert-{user_.username}.pem")
+
+            if isinstance(user_, user.IPAUser):
+                self.ipa_ca.request_cert(csr, user_.username, user_.cert)
+            else:
+                self.local_ca.request_cert(csr, user_.username, user_.cert)
+
+        user_.card.enroll()
+
+        dump_to_json(user_.card)
+        # Update the dump for current user to link it with the card
+        dump_to_json(user_)
+
+    def cleanup(self):
```

Comment:
What is the typical usage of this `cleanup` method? Removing CAs, so different CA could be created for a different test? I am wondering because a lot happens during setup and only CAs are cleaned up.